pkg/components/pluggable/discovery.go

```diff
@@ -0,0 +1,176 @@
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/dapr/dapr/utils"
+	"github.com/dapr/kit/logger"
+
+	"github.com/pkg/errors"
+
+	"github.com/jhump/protoreflect/grpcreflect"
+
+	"google.golang.org/grpc"
+	reflectpb "google.golang.org/grpc/reflection/grpc_reflection_v1alpha"
+)
+
+var (
+	discoveryLog        = logger.NewLogger("pluggable-components-discovery")
+	onServiceDiscovered map[string]func(name string, dialer GRPCConnectionDialer)
+)
+
+func init() {
+	onServiceDiscovered = make(map[string]func(name string, dialer GRPCConnectionDialer))
+}
+
+// AddServiceV1DiscoveryCallback adds a callback function that should be called when the given service was discovered.
+func AddServiceV1DiscoveryCallback(serviceName string, callbackFunc func(name string, dialer GRPCConnectionDialer)) {
+	onServiceDiscovered[withV1(serviceName)] = callbackFunc
+}
+
+// removeExt removes file extension
+func removeExt(fileName string) string {
+	return fileName[:len(fileName)-len(filepath.Ext(fileName))]
+}
+
+const (
+	SocketFolderEnvVar  = "DAPR_PLUGGABLE_COMPONENTS_SOCKETS_FOLDER"
+	defaultSocketFolder = "/dapr-components-sockets"
+)
+
+// GetSocketFolderPath returns the shared unix domain socket folder path
+func GetSocketFolderPath() string {
+	return utils.GetEnvOrElse(SocketFolderEnvVar, defaultSocketFolder)
+}
+
+const (
+	// the current components proto version
+	protoV1      = "v1"
+	protoPackage = "dapr.proto.components." + protoV1
+)
+
+// withV1 adds the service name using the v1 package.
+func withV1(serviceName string) string {
+	return fmt.Sprintf("%s.%s", protoPackage, serviceName)
+}
+
+type service struct {
+	// protoRef is the proto service name
+	protoRef string
+	// componentName is the component name that implements such service.
+	componentName string
+	// dialer is the used grpc connectiondialer.
+	dialer GRPCConnectionDialer
+}
+
+type reflectServiceClient interface {
+	ListServices() ([]string, error)
+}
+
+// serviceDiscovery returns all available discovered pluggable components services.
+// uses gRPC reflection package to list implemented services.
+func serviceDiscovery(reflectClientFactory func(string) (reflectServiceClient, *grpc.ClientConn, error)) ([]service, error) {
+	services := []service{}
+	componentsSocketPath := GetSocketFolderPath()
+	_, err := os.Stat(componentsSocketPath)
+
+	if os.IsNotExist(err) { // not exists is the same as empty.
+		return services, nil
+	}
+
+	log.Debugf("loading pluggable components under path %s", componentsSocketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := os.ReadDir(componentsSocketPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list pluggable components unix sockets")
+	}
+
+	for _, dirEntry := range files {
+		if dirEntry.IsDir() { // skip dirs
+			continue
+		}
+
+		f, err := dirEntry.Info()
+		if err != nil {
+			return nil, err
+		}
+
+		absPath := fmt.Sprintf("%s/%s", componentsSocketPath, f.Name())
+		if !utils.IsSocket(f) {
```

Comment:
Does this work on Windows too?

Reply:
Yes it should returns false, but indeed the go api for filesystem is very biased by unix-based OS

However, the entire pluggable components feature is designed to work only on unix-like systems. For instance, if I expect this to work on other OSs I should use os.PathSeparator rather than hardcoded “/“,

as a followup of your comment I’m going to add an explicit warning when trying to use pluggable components on windows, thanks

As a followup I’ll

Reply:
Ok perfect. I missed that this wasn't (yet?) designed for usage on Windows.

PS: `filepath.Join()` is normally better than manually dealing with `os.PathSeparator`

Reply:
>Ok perfect. I missed that this wasn't (yet?) designed for usage on Windows.

TBH I didn't discuss that with maintainers yet, but this was a design decision even before I joining

> PS: filepath.Join() is normally better than manually dealing with os.PathSeparator

fair enough